### PR TITLE
Transformer encoder API + get_model API + BERT

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##

This PR contains the following changes related to **transformer**, and they are ready for review:
- expose transformer encoder related APIs as public APIs
- include pre-trained transformer model in the get_model API
- refactor transformer building blocks,  such that they can be reused for BERT 

This PR also contains WIP changes for **BERT**, including:
- BERT building blocks
- TSVDataset and BERT data preprocessing script
- BERT fine-tuning script

### Essentials ###
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
@sxjscience @szhengac @leezu @cgraywang @szha

File: gluonnlp/model/bert.py [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+            prefix=prefix, params=params,
+            # extra configurations for the BERT-style transformer encoder cell
+            attention_use_bias=True, attention_proj_use_bias=True,
+            use_bert_layer_norm=True, use_bert_ffn=True)
+
+###############################################################################
+#                          FULL MODEL (WIP)                                   #
+###############################################################################
+
+class BERTModel(Block):
```
Comment (eric-haibin-lin) [inline gluonnlp/model/bert.py]: This class is still WIP

File: gluonnlp/model/seq2seq_encoder_decoder.py [hunk 2]
```diff
@@ -27,24 +27,6 @@
     MultiHeadAttentionCell
 
 
-def _list_bcast_where(F, mask, new_val_l, old_val_l):
```
Comment (eric-haibin-lin) [inline gluonnlp/model/seq2seq_encoder_decoder.py]: This function is not used anywhere.

File: gluonnlp/model/transformer.py [hunk 3]
```diff
@@ -17,8 +17,8 @@
 # specific language governing permissions and limitations
 # under the License.
 """Encoder and decoder usded in sequence-to-sequence learning."""
-__all__ = ['TransformerEncoder', 'TransformerDecoder',
-           'get_transformer_encoder_decoder', 'get_model']
+__all__ = ['TransformerEncoder', 'PositionwiseFFN', 'TransformerEncoderCell',
+           'transformer_en_de_512']
```
Comment (eric-haibin-lin) [inline gluonnlp/model/transformer.py]: Originally there was TransformerEncoder. Now there are 3 classes:
BaseTransformerEncoder
-> TransformerEncoder
-> BERTEncoder
where TransformerEncoder and BERTEncoder are just wrappers on top of BaseTransformerEncoder

File: gluonnlp/model/transformer.py [hunk 4]
```diff
@@ -210,138 +253,18 @@ def hybrid_forward(self, F, inputs, mask=None):  # pylint: disable=arguments-dif
         return outputs, additional_outputs
 
 
-class TransformerDecoderCell(HybridBlock):
```
Comment (eric-haibin-lin) [inline gluonnlp/model/transformer.py]: I made no changes to Decoders. They are not exposed as public APIs either. I simply moved it around.

File: gluonnlp/model/transformer.py [hunk 5]
```diff
@@ -522,6 +480,264 @@ def hybrid_forward(self, F, inputs, states=None, valid_length=None, position_wei
...
+            bias_initializer=bias_initializer, prefix=prefix, params=params,
+            # extra configurations for the original transformer
+            positional_weight='sinusoidal', use_bert_encoder=False,
+            use_layer_norm_before_dropout=True, scale_embed=True)
+
+###############################################################################
+#                                DECODER                                      #
+###############################################################################
+
+class TransformerDecoderCell(HybridBlock):
```
Comment (eric-haibin-lin) [inline gluonnlp/model/transformer.py]: Was simply moved around

File: gluonnlp/model/translation.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+import warnings
+import numpy as np
+from mxnet.gluon import Block
+from mxnet.gluon import nn
+import mxnet as mx
+from . import BeamSearchScorer, BeamSearchSampler
+
+
+class NMTModel(Block):
```
Comment (eric-haibin-lin) [inline gluonnlp/model/translation.py]: This was simply moved from scripts/machine_translation

File: scripts/bert/finetune_classifier.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+for name in tf_params:
+    if name not in loaded:
+        print('not loading', name)
+
+loss_function = gluon.loss.SoftmaxCELoss()
+loss_function.hybridize(static_alloc=static_alloc)
+
+metric = mx.metric.Accuracy()
+
+trans = SentenceClassificationTrans(tokenizer, MRPCDataset.get_labels(), args.max_len)
```
Comment (eric-haibin-lin) [inline scripts/bert/finetune_classifier.py]: TODO: need use custom optimizer and initializer

File: scripts/bert/tokenization.py [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+from __future__ import division
+from __future__ import print_function
+
+import collections
+import unicodedata
+import six
+import io
+
+
+def convert_to_unicode(text):
```
Comment (eric-haibin-lin) [inline scripts/bert/tokenization.py]: Simply moved from the google repo.

File: src/gluonnlp/data/dataset.py [hunk 9]
```diff
@@ -78,6 +78,79 @@ def _corpus_dataset_process(s, bos, eos):
...
+    field_separator : function or None, default Splitter('\t')
+        A function that splits each sample string into list of text fields.
+        If None, raw samples are returned according to `sample_splitter`.
+    num_discard_samples : int, default 0
+        Number of samples discarded at the head of the first file.
+    field_indices : list of int, default None
+        If set, for each sample, only fields with provided indices are selected as the output.
+        Otherwise all fields are returned.
+
+    Example::
```
Comment (szha) [inline src/gluonnlp/data/dataset.py]: I'm enabling doctest in #402.

Comment (eric-haibin-lin) [reply src/gluonnlp/data/dataset.py]: The api takes a path to the file as the input. To run this we have to create a file. Showing the code for file creation is verbose... Maybe we don't want to enforce all examples to be verified by doctest. wdyt

Comment (szha) [reply src/gluonnlp/data/dataset.py]: You can either check a small file in, or have it hosted and downloaded. I'm doing the same for sentencepiece tokenizer where the tokenizer requires a bpe model file

File: src/gluonnlp/model/transformer.py [hunk 10]
```diff
@@ -522,6 +480,264 @@ def hybrid_forward(self, F, inputs, states=None, valid_length=None, position_wei
...
+
+class TransformerDecoderCell(HybridBlock):
+    """Structure of the Transformer Decoder Cell.
+
+    Parameters
+    ----------
+    attention_cell : AttentionCell or str, default 'multi_head'
+        Arguments of the attention cell.
+        Can be 'multi_head', 'scaled_luong', 'scaled_dot', 'dot', 'cosine', 'normed_mlp', 'mlp'
+    units : int
```
Comment (sxjscience) [inline src/gluonnlp/model/transformer.py]: We'd better explain `units`.

File: scripts/bert/bert.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+                self.classifier.add(nn.Dropout(rate=dropout))
+            self.classifier.add(nn.Dense(units=num_classes, flatten=False))
+
+    def forward(self, inputs, token_types, valid_length=None):
+        """Generate the unnormalized score for the given the input sequences.
+
+        Parameters
+        ----------
+        inputs : NDArray
+        token_types : NDArray
```
Comment (sxjscience) [inline scripts/bert/bert.py]: Better explain the `token_types`.

File: src/gluonnlp/data/dataset.py (hunk 9 above)
Comment (sxjscience) [reply src/gluonnlp/data/dataset.py]: +1 for doctesting.

File: src/gluonnlp/data/dataset.py [hunk 12]
```diff
@@ -78,6 +78,79 @@ def _corpus_dataset_process(s, bos, eos):
...
+        if not self._field_indices:
+            return fields
+        return [fields[i] for i in self._field_indices]
+
+    def _read(self):
+        all_samples = []
+        for filename in self._filenames:
+            with io.open(filename, 'r', encoding=self._encoding) as fin:
+                content = fin.read()
+            samples = (s for s in self._sample_splitter(content) if not self._discard(s))
```
[the remark this answers is not in the document]
Comment (sxjscience) [reply src/gluonnlp/data/dataset.py]: A for-loop would be easier to read.

File: scripts/bert/dataset.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+      tokens_a.pop()
+    else:
+      tokens_b.pop()
+
+class SentenceClassificationTrans(object):
+    """Dataset Transformation for BERT-style Sentence Classification.
+
+    Parameters
+    ----------
+    tokenizer : BasicTokenizer or FullTokensize.
```
Comment (sxjscience) [inline scripts/bert/dataset.py]: FullTokensize should be FullTokenizer

File: scripts/bert/dataset.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+    total_length = len(tokens_a) + len(tokens_b)
+    if total_length <= max_length:
+      break
+    if len(tokens_a) > len(tokens_b):
+      tokens_a.pop()
+    else:
+      tokens_b.pop()
+
+class SentenceClassificationTrans(object):
+    """Dataset Transformation for BERT-style Sentence Classification.
```
Comment (sxjscience) [inline scripts/bert/dataset.py]: We need an example.

File: scripts/bert/finetune_classifier.py [hunk 15]
```diff
@@ -0,0 +1,212 @@
+"""
+Bidirectional Encoder Representations from Transformers.
+
+=================================
+
+This example shows how to implement the Transformer model with Gluon NLP Toolkit.
```
Comment (sxjscience) [inline scripts/bert/finetune_classifier.py]: According to my understanding, this example is about train the BERT classifier.

File: src/gluonnlp/model/bert.py [hunk 16]
```diff
@@ -0,0 +1,596 @@
...
+    word_embed : Block or None, default None
+        The word embedding. If set to None, word_embed will be constructed using embed_size and
+        embed_dropout.
+    token_type_embed : Block or None, default None
+        The token type embedding. If set to None and the token_type_embed will be constructed using
+        embed_size and embed_dropout.
+    use_pooler : bool, default True
+        Whether to include the pooler which converts the encoded sequence tensor of shape
+        (batch_size, seq_length, units) to a tensor of shape (batch_size, units)
+        for for segment level classification task.
```
Comment (sxjscience) [inline src/gluonnlp/model/bert.py]: Small typo. "for for" should be "for"

File: src/gluonnlp/model/seq2seq_encoder_decoder.py [hunk 17]
```diff
@@ -17,7 +17,7 @@
 # specific language governing permissions and limitations
 # under the License.
 """Encoder and decoder usded in sequence-to-sequence learning."""
-__all__ = ['Seq2SeqEncoder', 'Seq2SeqDecoder']
+__all__ = ['Seq2SeqEncoder']
```
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply src/gluonnlp/model/seq2seq_encoder_decoder.py]: I intentionally didn't expose all the decoder classes since there's no conclusion on the decoder APIs yet. we can revisit that when there's bandwidth

File: src/gluonnlp/model/transformer.py [hunk 18]
```diff
@@ -17,8 +17,8 @@
 # specific language governing permissions and limitations
 # under the License.
 """Encoder and decoder usded in sequence-to-sequence learning."""
-__all__ = ['TransformerEncoder', 'TransformerDecoder',
-           'get_transformer_encoder_decoder', 'get_model']
+__all__ = ['TransformerEncoder', 'PositionwiseFFN', 'TransformerEncoderCell',
```
[the remark this answers is not in the document]
Comment (eric-haibin-lin) [reply src/gluonnlp/model/transformer.py]: See https://github.com/dmlc/gluon-nlp/pull/409#discussion_r234534772

File: scripts/bert/tokenization.py [hunk 19]
```diff
@@ -0,0 +1,310 @@
...
+      if cp == 0 or cp == 0xfffd or _is_control(char):
+        continue
+      if _is_whitespace(char):
+        output.append(" ")
+      else:
+        output.append(char)
+    return "".join(output)
+
+
+class WordpieceTokenizer(object):
```
Comment (szhengac) [inline scripts/bert/tokenization.py]: Detokenizer is also needed to for bleu evaluation.

File: src/gluonnlp/model/translation.py [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+                    self.tgt_embed = nn.HybridSequential(prefix='tgt_embed_')
+                    with self.tgt_embed.name_scope():
+                        self.tgt_embed.add(
+                            nn.Embedding(input_dim=len(tgt_vocab), output_dim=embed_size,
+                                         weight_initializer=embed_initializer))
+                        self.tgt_embed.add(nn.Dropout(rate=embed_dropout))
+        # Construct tgt proj
+        if tie_weights:
+            self.tgt_proj = nn.Dense(units=len(tgt_vocab), flatten=False,
+                                     params=self.tgt_embed.params, prefix='tgt_proj_')
```
Comment (szhengac) [inline src/gluonnlp/model/translation.py]: It has been changed on #413.

File: src/gluonnlp/model/transformer.py [hunk 21]
```diff
@@ -738,9 +990,9 @@ def forward(self, step_input, states, mask=None):  #pylint: disable=arguments-di
                                         axis=0, size=step_input.shape[0])
         steps = mx.nd.arange(step_input.shape[1], ctx=step_input.context)
         states.append(steps)
+        scaled_step_input = step_input * math.sqrt(step_input.shape[-1])
```
Comment (szhengac) [inline src/gluonnlp/model/transformer.py]: To be consistent with Encoder. if self._scale_embed: is missing.

**Comment by szhengac** [comment]
I would recommend splitting this pr into multiple ones. The current pr looks too cumbersome to maintain.

File: src/gluonnlp/data/utils.py [hunk 22]
```diff
@@ -352,7 +358,7 @@ def line_splitter(s):
     return s.splitlines()
 
 def whitespace_splitter(s):
```
Comment (leezu) [inline src/gluonnlp/data/utils.py]: This can probably be removed (superseded by Splitter)? Only reason for introducing it was that `lambda s: s.split()` can not be pickled and does not work in a multiprocessing data pipeline

File: scripts/bert/dataset.py (hunk 14 above)
Comment (eric-haibin-lin) [reply scripts/bert/dataset.py]: Added to `__call__`

File: scripts/bert/tokenization.py (hunk 19 above)
Comment (eric-haibin-lin) [reply scripts/bert/tokenization.py]: Yes. I'll track that in #425

File: src/gluonnlp/data/utils.py (hunk 22 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/data/utils.py]: I'll track this separately in #425

File: src/gluonnlp/model/transformer.py (hunk 21 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/model/transformer.py]: Added.

File: src/gluonnlp/model/translation.py (hunk 20 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/model/translation.py]: Thanks for pointing this out!

**Comment by sxjscience** [comment]
LGTM. Currently there's some problems with the doctest. I'll approve once that's fixed.

File: src/gluonnlp/model/bert.py [hunk 23]
```diff
@@ -0,0 +1,609 @@
...
+import mxnet as mx
+from .transformer import BasePositionwiseFFN, BaseTransformerEncoderCell, BaseTransformerEncoder
+from .block import GELU
+from .utils import _load_vocab, _load_pretrained_params
+
+###############################################################################
+#                              COMPONENTS                                     #
+###############################################################################
+
+class BERTLayerNorm(HybridBlock):
```
Comment (szhengac) [inline src/gluonnlp/model/bert.py]: What is the difference from the Layernorm? The way of adding epsilon?

File: src/gluonnlp/model/bert.py [hunk 24]
```diff
@@ -0,0 +1,609 @@
...
+        # encoding
+        outputs, additional_outputs = self.encoder(embedding, None, valid_length)
+        return outputs, additional_outputs
+
+    def _apply_pooling(self, sequence):
+        """Generate the representation given the inputs.
+
+        This is used for pre-training or fine-tuning a BERT model.
+        """
+        outputs = sequence.slice(begin=(None, 0, None), end=(None, 1, None)).squeeze(axis=1)
```
Comment (szhengac) [inline src/gluonnlp/model/bert.py]: We can just use sequence[:, 0, :].

File: src/gluonnlp/model/bert.py [hunk 25]
```diff
@@ -0,0 +1,609 @@
...
+        return self.pooler(outputs)
+
+    def _decode(self, sequence, valid_length=None):
+        """Generate unormalized prediction for the masked language model task.
+
+        This is only used for pre-training the BERT model.
+        """
+        if valid_length is None:
+            last_step = sequence.slice(begin=(None, -1, None), end=(None, None, None))
```
Comment (szhengac) [inline src/gluonnlp/model/bert.py]: Similar, sequence[:, -1, :].

**Comment by szhengac** [comment]
Also, we need to make sure that the existing pretrained model can be loaded and perform the testing correctly.

**Comment by mli** [comment]
Job PR-409/22 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-409/22/index.html

**Comment by eric-haibin-lin** [comment]
@szhengac Yes. I loaded the same model in the transformer notebook for inference and got exactly the same loss and bleu score.

**Comment by mli** [comment]
Job PR-409/27 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-409/27/index.html
